trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+            item_representation = Generator(seed=seed).binomial(
+                n=1, p=0.5, size=(num_attributes, num_items)
+            )
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
+        self.all_interactions = self.all_interactions.append(interaction_df, ignore_index=True)
+
+    def update_predicted_scores(self):
+        """
+        Calculates predicted scores for every user-item pair by first running an
+        implicit MF algorithm on the interaction data. A glorified wrapper function,
+        but necessary to keep the functionality in BaseRecommender working.
+        """
+        # if no interactions have happened yet, do a simple dot product
+        if self.all_interactions.size == 0:
+            super().update_predicted_scores()
+        else:
+            self.fit_mf()
+
+    def run(
+        self,
+        timesteps=50,
+        startup=False,
+        train_between_steps=False,
+        random_items_per_iter=0,
+        vary_random_items_per_iter=False,
+        repeated_items=True,
+        no_new_items=False,
+    ):
+        """
+        Just a simple wrapper so that by default, the RS does not refit the ImplicitMF model
+        at every timestep of the simulation. See the `run` method in
+        :`~models.recommender.BaseRecommender` for more details about what this does.
+        """
+        # reset interactions tracker so that interactions captured
+        # are only for the duration of this particular run
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])
+        super().run(
+            timesteps,
+            startup,
+            train_between_steps,
+            random_items_per_iter,
+            vary_random_items_per_iter,
+            repeated_items,
+            no_new_items=no_new_items,
+        )
+
+    def startup_and_train(self, timesteps=50, no_new_items=True):
```

Comment:
I feel like calling this function (maybe generally throughout the codebase) `startup_and_train` a bit misleading since in at least the MF case, it doesn't actually train anything unless you call `mf_fit`. Are there other models where some sort of training happens within `startup_and_train`?

Reply:
See previous comment, but this method does call `fit_mf()` because the super class's `startup_and_train()` calls `update_predicted_scores`, which then calls `fit_mf()`. 

All other models in our library have this same structure, where at the end of `startup_and_train()`, data used from the startup iterations is then used to re-calculate the predicted user-item scores, which is basically the same idea as "training", because that's what actually changes the recommendations for each user.

Reply:
Yep. I see that now. Not sure why I continue to have so much difficulty figuring this out. I'll get it eventually!

Reply:
It is confusing! Haha, I just have the benefit of being immersed in the guts of the library for too long 😛